### ISSUE TITLE
sendToolListChanged drops notifications on stateless Streamable HTTP transports (missing relatedRequestId)

### ISSUE THREAD

**Issue by luizgribeiro** [modelcontextprotocol/typescript-sdk#2232]
## Summary

When `tools/list_changed` is triggered from within a tool handler (e.g., via `enable()`, `disable()`, or `update()`), the notification is silently dropped on stateless Streamable HTTP transports because it lacks a `relatedRequestId`.

## Reproduction

1. Create an MCP server using `StreamableHTTPServerTransport` with `sessionIdGenerator: undefined` (stateless)
2. Register a tool that, when called, enables another tool via the `RegisteredTool.enable()` handle
3. Call that tool from a client
4. The `tools/list_changed` notification is never delivered to the client

## Root cause

`RegisteredTool.enable()` calls `update({ enabled: true })` which calls `sendToolListChanged()`:

```js
// mcp.js — _createRegisteredTool
update: updates => {
  // ...
  registeredTool.enabled = updates.enabled;
  this.sendToolListChanged(); // ← no request context
}
```

`sendToolListChanged()` calls `this.server.notification(...)` without options:

```js
// server/index.js
async sendToolListChanged() {
  return this.notification({ method: 'notifications/tools/list_changed' });
  //                         ^ no relatedRequestId
}
```

In the transport's `send()` method, messages without `relatedRequestId` are routed exclusively to the standalone GET SSE stream. On stateless transports, no such stream exists, so the notification is silently dropped.

This contradicts the spec, which says:

> The server **MAY** send JSON-RPC *requests* and *notifications* before sending a JSON-RPC *response*. These messages **SHOULD** relate to the originating client *request*.

— [Streamable HTTP transport spec (2025-03-26)](https://modelcontextprotocol.io/specification/2025-03-26/basic/transports#streamable-http)

## Workaround

The SDK already provides `extra.sendNotification()` in tool handler callbacks, which correctly sets `relatedRequestId`. We work around the issue by:

1. Setting `.enabled = true` directly on `_registeredTools[name]` (bypassing `enable()` to avoid the broken notification)
2. Sending `tools/list_changed` ourselves via `extra.sendNotification({ method: 'notifications/tools/list_changed' })`

This routes the notification onto the POST response SSE stream, making it work on both stateful and stateless transports.

## Suggested fix

When `enable()`, `disable()`, or `update()` are called from within a request handler, the resulting `sendToolListChanged()` should include the `relatedRequestId` of the in-flight request. This would route the notification to the POST response stream per the spec.

One approach: `sendToolListChanged` (and the other `send*ListChanged` methods) could accept an optional `options` parameter that gets forwarded to `notification()`:

```js
async sendToolListChanged(options) {
  return this.notification({ method: 'notifications/tools/list_changed' }, options);
}
```

And `_createRegisteredTool`'s `update()` could accept and forward those options, or the `RegisteredTool` handle could expose a way to pass request context.

Alternatively, the SDK could use `AsyncLocalStorage` to automatically capture the `relatedRequestId` when a tool handler is executing, so existing `enable()`/`disable()` calls work without changes.

## Environment

- `@modelcontextprotocol/sdk`: 1.29.0
- Transport: `StreamableHTTPServerTransport` (stateless, `sessionIdGenerator: undefined`)
- Affects: `sendToolListChanged`, `sendResourceListChanged`, `sendPromptListChanged` — all `*ListChanged` notifications have the same issue

**Comment by Brandon1138** [comment]
I can take this one. I've reproduced the issue on the v1.x line and I'm working through a targeted fix with regression coverage for stateless streamable HTTP

**Comment by Brandon1138** [comment]
I have a fix ready, but GitHub is preventing me from opening a PR because this repository currently limits PR creation to collaborators.

Branch: https://github.com/Brandon1138/typescript-sdk/tree/fix/issue-2232-list-changed-request-context

Compare: https://github.com/modelcontextprotocol/typescript-sdk/compare/v1.x...Brandon1138:typescript-sdk:fix/issue-2232-list-changed-request-context

The diff is based on `v1.x` and currently shows 1 commit / 3 files changed, mergeable against `v1.x`.

Happy to open the PR if permissions are enabled.

**Comment by norika1207-lab** [comment]
This looks worth pinning down as a protocol-level regression rather than only an implementation bug.

For `sendToolListChanged drops notifications on stateless Streamable HTTP transports`, the acceptance test I would want is a minimal client/server fixture that records the exact JSON-RPC or HTTP exchange, then asserts both the wire result and the local lifecycle state after the failure path. In MCP deployments, the dangerous failures are often not just wrong return values; they are pending requests, dropped notifications, mismatched session state, or errors that get translated into a generic response and become impossible to diagnose.

A useful fix boundary would preserve three signals: the protocol error sent to the peer, the local exception or cancellation reason available to application code, and a trace/log field that ties the two together. That makes the behavior debuggable for agent hosts without requiring every host to special-case this SDK detail.

**Comment by Brandon1138** [comment]
Makes sense. I’ll add a protocol-level stateless Streamable HTTP fixture that records the raw POST/SSE exchange and asserts that the `list_changed` notification and `tools/call` result are delivered on the same response stream, alongside the local lifecycle assertions.

I’ll scope this to the acceptance test unless it exposes a gap in the implementation.

**Comment by Brandon1138** [comment]
Added the protocol-level acceptance coverage.

Branch:
https://github.com/Brandon1138/typescript-sdk/tree/fix/issue-2232-list-changed-request-context

Compare:
https://github.com/modelcontextprotocol/typescript-sdk/compare/v1.x...Brandon1138:typescript-sdk:fix/issue-2232-list-changed-request-context

The new raw HTTP fixture asserts that stateless Streamable HTTP delivers both `notifications/tools/list_changed`
and the `tools/call` result on the same POST/SSE response stream. The dynamic scenario covers the local lifecycle/
scoping behavior for related, unrelated, and detached notifications.

Validation:
 - `npm run check`
 - `npm run build`
 - targeted dynamic and hosting e2e regressions
 - `git diff --check upstream/v1.x`

**Comment by Srinu0342** [comment]
Opened a fix for this: [#2522](https://github.com/modelcontextprotocol/typescript-sdk/pull/2522)

Root cause matches what's been described in the thread — Server.sendToolListChanged()/sendResourceListChanged()/sendPromptListChanged() send with no relatedRequestId, so the Streamable HTTP transport always routes them at the standalone GET SSE stream. On a stateless transport that stream never exists, so a notification fired from inside a request handler (e.g. RegisteredTool.enable() called mid-tools/call) is silently dropped, even though the handler's own request has a live response stream right there.

The fix threads the currently-executing request's id through implicitly: Protocol runs each request handler inside a tracked context (_runHandlerInContext), and when a notification is sent without an explicit relatedRequestId, it now defaults to whichever request is currently in flight (_currentInflightRequestId) instead of leaving it unset. A caller-supplied relatedRequestId still always wins, and nothing about the public API changes — sendToolListChanged() etc. still take no arguments. Server backs the two hooks with AsyncLocalStorage via the existing per-runtime _shims subpath (Node gets the real thing; workerd/browser get a single-slot, synchronous-scope-only fallback, documented as such, since node:async_hooks isn't available there).

Added a regression test (packages/server/test/server/statelessListChangedRelatedRequestId.test.ts) that registers a disabled tool + a tool that enables it from inside its own handler, drives the exchange through a stateless WebStandardStreamableHTTPServerTransport, and asserts the response stream carries both the tools/call result and notifications/tools/list_changed. Red before the fix, green after. Full suite (server/core-internal/client/server-legacy) stays green, typecheck/lint/build clean.
